nomad/plan_apply.go

```diff
@@ -88,18 +100,29 @@ func (p *planner) planApply() {
 			return
 		}
 
-		// Check if out last plan has completed
+		// If last plan has completed get a new snapshot
 		select {
-		case <-waitCh:
-			waitCh = nil
+		case idx := <-planIndexCh:
+			// Previous plan committed. Discard snapshot and ensure
+			// future snapshots include this plan.
+			prevPlanResultIndex = max(prevPlanResultIndex, idx)
+			planIndexCh = nil
 			snap = nil
 		default:
 		}
 
+		if snap != nil {
+			// If snapshot isn't new enough, discard it
+			minIndex := max(prevPlanResultIndex, pending.plan.SnapshotIndex)
+			if idx, err := snap.LatestIndex(); err != nil || idx < minIndex {
+				snap = nil
+			}
+		}
+
 		// Snapshot the state so that we have a consistent view of the world
 		// if no snapshot is available
-		if waitCh == nil || snap == nil {
-			snap, err = p.fsm.State().Snapshot()
+		if planIndexCh == nil || snap == nil {
```

Comment:
I feel like I need to build a SAT-solver to understand some of the looping and branching here :P, and given that oder is critical and we cannot substitute them with `else if`, it might be nice explaining the high-level flow in a comment, e.g. why does ch/snap/etc nilness signify in each of the conditionals.